feat: allow single quotes in text selector engine

mxschmitt opened the pull request:
This change also allows the usage of single quotes in the text selector engine. Also it fixes a bug which was persistent when you tried to use `text="` which lead to a stack trace that JSON could not be parsed.

Not sure where to put the `hasTextSelectorSurroundings` helper function since its needed in two locations, but `src/injected/textSelectorEngine.ts` has no access to the other one, thats why the test is failing.

Closes #1882

mxschmitt committed:
feat: allow single quotes in text selector engine

dgozman commented on src/selectors.ts:
```diff
@@ -184,7 +185,7 @@ export class Selectors {
         name = part.substring(0, eqIndex).trim();
         body = part.substring(eqIndex + 1);
-      } else if (part.startsWith('"')) {
+      } else if (hasTextSelectorSurroundings(part)) {
```
Unfortunately, we do not have a way to share the code between injected (in the page) and selectors (in the node). I'd just copy it here.

Note that you changing semantics from 'starts with' to 'starts and ends with' - this should be reflected in the docs.

dgozman commented on src/selectors.ts:
```diff
@@ -184,7 +185,7 @@ export class Selectors {
-      } else if (part.startsWith('"')) {
+      } else if (hasTextSelectorSurroundings(part)) {
         name = 'text';
         body = part;
```
Now that you brought unpleasant JSON.parse error, perhaps we should try to JSON.parse right here in node to get a nice stack trace and/or error message? I'd also add a test for that.

dgozman reviewed (commented):
This looks very good!

mxschmitt committed:
feat: added handling if text selector itself contains quotes

mxschmitt commented:
Added the documentation changes and the enhancement, so that we'll parse the input as JSON. I thought also about the edge case like that:

```js
 await page.setContent(`<div>"ya</div>`);
 expect(await page.$eval(`'"ya'`, e => e.outerHTML)).toBe('<div>"ya</div>');
```

That would end up in a JSON parsing error, thats why I've added a check for `innerText.includes('\\')`. And if it's true, then parse it as JSON. Maybe there's a better way to handle that.

Or do you think that this is too uncommon and we should remove it?

dgozman commented on src/injected/textSelectorEngine.ts:
```diff
@@ -48,17 +48,27 @@ export function createTextSelector(shadow: boolean): SelectorEngine {
+  if (hasTextSelectorSurroundings(selector)) {
+    const innerText = selector.slice(1, selector.length - 1);
+    // Use JSON.parse since we want to parse escaped characters
+    const parsed = innerText.includes('\\') ? JSON.parse(`"${innerText}"`) : innerText;
```
This does not work for `text='foo"bar'` because it will try to parse `"foo"bar"`. I guess we have to do our own parser, or forget about single quotes. WDYT?

mxschmitt replied:
From my side to leave out the single quote support would be not that good. Too often there are code bases / users out there which would run into that issue.

I'm just thinking, why at all do we want to allow inside the text selector JSON based input?

A few test cases which I've tested and come to the conclusion, that we might not need `JSON.parse`:

- `text="foo"yes"very'tricky"` - would work since we do only test directly for the text inside the quotes
- `text=foo\\backslash` - would test for `foo\backslash`
- `text=""ya"` - would test for `"ya"`

I see only the `text="\\"ya"` case in the [tests](https://github.com/mxschmitt/playwright/blob/feature/enh-text-selector-engine/test/queryselector.spec.js#L516) which makes JSON parse useful. But since we are adding single quote support and dont use JSON parse anymore, they can either do `text=""ya"` or `text='"ya'` to handle that specific use case which should also be very rare.

dgozman replied:
The main reason is to follow standard practices with strings. If you quote a string, you should escape the quote inside, so that we can figure out where does the string end. For example, consider `text="foo">>text="bar"`: is it a single text selector with `foo">>text="bar` body, or two chained text selectors?

mxschmitt replied:
hmm, good point. Should it be allowed to have multiple text selectors in a single selector?

If yes, then we probably have to write our own parser. If not, I see not a problem, that `text="foo">>text="bar"` will lead to `foo">>text="bar`.

Maybe in the future we should also allow `#foo > text="foo"`, but that would be out of the scope of this PR.

dgozman commented on src/selectors.ts:
```diff
@@ -184,9 +189,10 @@ export class Selectors {
         name = 'text';
-        body = part;
+        const innerText = part.slice(1, part.length - 1);
+        body = innerText.includes('\\') ? JSON.parse(`"${innerText}"`) : innerText;
```
Note this is not exactly equivalent, since it turns strict notation to the relaxed one by getting rid of the quotes. I was thinking about trying the JSON.parse to only check for the error earlier.

dgozman commented on docs/selectors.md:
```diff
@@ -109,11 +109,11 @@ Note that `xpath` does not pierce shadow roots.
 - Input elements of the type `button` and `submit` are rendered with their value as text, and text engine finds them. For example, `text=Login` matches `<input type=button value="Login">`.
 
-Malformed selector starting with `"` is assumed to be a text selector. For example, Playwright converts `page.click('"Login"')` to `page.click('text="Login"')`.
+Malformed selector surrounded with quotes (`"` or `'`) is assumed to be a text selector. For example, Playwright converts `page.click('"Login"')` to `page.click('text="Login"')`.
```
Unfortunately, we have this in multiple places now. I guess searching for "starting with" or "malformed" could help.

mxschmitt replied:
Will go over it once its fully clarified 👍

dgozman reviewed (commented):
This turned out to be harder than I thought...

mxschmitt commented:
Reworked by #1952
